server/workflow/workflow_server.go

```diff
@@ -204,6 +205,15 @@ func (s *workflowServer) WatchWorkflows(req *workflowpkg.WatchWorkflowsRequest,
 	log.Debug("Piping events to channel")
 	defer log.Debug("Result channel done")
 
+	// Eagerly send the headers so that we can begin our keepalive loop if no results are received
+	// immediately.  Without this, we cannot detect a streaming response, and we can't write to the
+	// response since a subsequent write by the stream causes an error.
+	err = ws.SendHeader(metadata.MD{})
```

Comment:
Would definitely appreciate extra scrutiny on this line - I am not familiar enough with gRPC to know the potential implications of this, though I did test that it seems to continue functioning the same.

Without this, the headers that are set are not sent until the first message is received, however that message could be very delayed if you're on a relatively inactive installation of Argo.  This should force the headers to be flushed to gRPC which then results in the headers being sent to the browser, allowing us to peek at the response to confirm that it's a streaming response and begin our keepalive loop.